API.md

```diff
@@ -0,0 +1,93 @@
+# Introduction
+This document provides API specification of NodeResourceTopology API which is a CRD
+based API that describes the structure of topology on a node. One of the goals when
+designing this API was to ensure hardware independent data structure, which allows
+us to have a future-proof API that represents hardware of varying internal topology.
+
+CRD instances must be created external to kubelet and independently. Each instance
+will be bound to one working node.
+
+# NodeTopology Object Definition
+
+```
+// NodeResourceTopology describes node resources and their topology.
+type NodeResourceTopology struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Zones      ZoneList      `json:"zones"`
+	Attributes AttributeList `json:"attributes,omitempty"`
+}
+
+// Zone represents a resource topology zone, e.g. socket, node, die or core.
+type Zone struct {
+	Name       string           `json:"name" protobuf:"bytes,1,opt,name=name"`
+	Type       string           `json:"type" protobuf:"bytes,2,opt,name=type"`
+	Parent     string           `json:"parent,omitempty" protobuf:"bytes,3,opt,name=parent"`
+	Costs      CostList         `json:"costs,omitempty" protobuf:"bytes,4,rep,name=costs"`
```

Comment:
Is cost for a zone enough for resources? Cost of cpus in different NUMA nodes can be different with gpus in different NUMA nodes because of the NVLink or NVSwitch.

Reply:
I think resources with special interconnects between them should then be modeled as separate zones, if you want to represent that in detail

Reply:
> I think resources with special interconnects between them should then be modeled as separate zones, if you want to represent that in detail

+1 from my side, perhaps we should "just" note in the docs (= API comments) that this is how envision the API to be used

Reply:
Good idea, will try to capture this in the docs.

Reply:
Do you mean we can have a zoneList like this?
{{NUMANode-gpu-0,cost=10}, {PCIE-0-gpu1, type=gpu, parent=NUMANode-gpu-0},{PCIE-0-gpu2, type=gpu, parent=NUMANode-gpu-0},{NUMANode-cpu-0,cost=20},{PCIE-0-cpu1, type=cpu, parent=NUMANode-cpu-0},{PCIE-0-cpu2, type=cpu, parent=NUMANode-cpu-0},}

Reply:
Yes, this is close but the thing to note here is that the cost is a list that represents distance of that zone from other zones and we represent this with a key-value pair (key capturing the zone and value capturing the cost) so in the example above cost should be indicative of distance of NUMANode-gpu-0 from NUMANode-cpu-0 in both the cases and therefore should be identical. But if there was another zone in the example we could have a different cost from these zones. Something like below:

```
{{NUMANode-gpu-0,cost=({NUMANode-cpu-0:10},{NUMANode-gpu-1:20}}}, {PCIE-0-gpu1, type=gpu, parent=NUMANode-gpu-0},{PCIE-0-gpu2, type=gpu, parent=NUMANode-gpu-0},
{NUMANode-cpu-0,cost=({NUMANode-gpu-0:10},{NUMANode-gpu-1:30}},{PCIE-0-cpu1, type=cpu, parent=NUMANode-cpu-0},{PCIE-0-cpu2, type=cpu, parent=NUMANode-cpu-0},
{NUMANode-gpu-1,cost=({NUMANode-cpu-0:30},{NUMANode-gpu-0:20}}, {PCIE-1-gpu1, type=gpu, parent=NUMANode-gpu-1},{PCIE-1-gpu2, type=gpu, parent=NUMANode-gpu-1}}
```

Hope this helps.

Reply:
> Good idea, will try to capture this in the docs.

thanks! also providing examples of topologies and their modellization would help.
I guess we will also need  somehow to address the point that there are multiple legal and equivalent modelization of the same hardware, not just one. Unlikely a blocker for alpha, though.

Reply:
Thanks for replying. I have another question. When I tried to write a sample for the case, I found that we have to add type in zoneName to avoid create same name for different type of device in one node or other level, do we still need a "type" property? Maybe type can be an attribute. HDYT? @swatisehgal

Reply:
And if I want to define a gpu in socket 0, numa 0, pcie 0. I have to define a list with 3 items. 
Maybe I want to write in this format: 
```
{Name: gpu-0, Attributes={Socket=0, Numa=0, PCIE=0}}
```
In what cases we need to care about the parent of differnet zones?

Reply:
We were aiming for an API that provides flexibility and has a hierarchical structure to represent relationship and proximity between various topological zones. It sounds like you were expecting a more flatter representation which is achievable with the current API we have. You can define all the resources within a single zone and use the attributes field like you are suggesting.

It is important to keep [this](https://github.com/kubernetes-sigs/noderesourcetopology-api/pull/1#discussion_r1453326813) comment from @ffromani in mind. We wanted to represent a wide range of hardware and resources and can only provide our recommendations on how the API can be _potentially_ used. But, there is no right and wrong here, if you would like it to use differently you can totally do that!

As long as the producer (exporter that creates instances of object) and the consumer (which could be a human or a software entity) have a shared understanding of the representation we should be okay. 

> In what cases we need to care about the parent of different zones?

A simple example would to be represent hardware differently where there are multiple NUMA nodes per socket vs multiple sockets per NUMA nodes. I would suggest that you refer to [this](https://docs.google.com/document/d/12kj3fK8boNuPNqob6F_pPU9ZTaNEnPGaXEooW1Cilwg/edit?usp=sharing) document for the rationale behind the design of this API.

Reply:
> Thanks for replying. I have another question. When I tried to write a sample for the case, I found that we have to add type in zoneName to avoid create same name for different type of device in one node or other level, do we still need a "type" property? Maybe type can be an attribute. HDYT? @swatisehgal 

For the cases we have in mind, we need the "type" property but it sounds like you would like it to be an optional field?